unit/memory-analyzer/gdb_api.cpp

```diff
@@ -18,13 +18,61 @@ Author: Malte Mues <mail.mues@gmail.com>
 #include <iostream>
 
 #include <memory-analyzer/gdb_api.cpp>
+
 #include <util/run.h>
+#include <util/tempfile.h>
 
-void compile_test_file()
+struct compile_test_filet
 {
-  REQUIRE(
-    run("gcc", {"gcc", "-g", "-o", "test", "memory-analyzer/test.c"}) == 0);
-}
+  compile_test_filet() : compiled("test", "")
+  {
+    temporary_filet tmp("test", ".c");
+    std::ofstream of(tmp().c_str());
+    REQUIRE(of.is_open());
+
+    of << "int x;\n"
+          "float y;\n"
+          "char z;\n"
+          "\n"
+          "char *s = \"abc\";\n"
+          "int *p;\n"
+          "void *vp;\n"
+          "int *np = 0;\n"
+          "void *vp_string;\n"
+          "\n"
+          "void checkpoint()\n"
+          "{\n"
+          "}\n"
+          "void checkpoint2()\n"
+          "{\n"
+          "}\n"
+          "\n"
+          "void func()\n"
+          "{\n"
+          "  checkpoint2();\n"
+          "}\n"
+          "\n"
+          "int main()\n"
+          "{\n"
+          "  x = 8;\n"
+          "  y = 2.5;\n"
+          "  z = 'c';\n"
+          "\n"
+          "  p = &x;\n"
+          "  vp = (void *)&x;\n"
+          "  vp_string = s;\n"
+          "\n"
+          "  checkpoint();\n"
+          "\n"
+          "  return 0;\n"
+          "}\n";
+    of.close();
+
+    REQUIRE(run("gcc", {"gcc", "-g", "-o", compiled(), tmp()}) == 0);
+  }
+
+  temporary_filet compiled;
+};
```

Comment:
This seems unnecessarily ugly, is there a reason to do it this way?

Reply:
If, as was the case before, we keep the contents in a separate file, then this file must either be included in the compiled unit test binary (e.g., via `#include`), or else the compiled unit test binary necessarily relies on the source tree being present as well. It seems reasonable to assume that the source tree will be present, but then also the exact location of the file must be known.

I made these changes as I found the above conditions not to be satisfied in my ad-hoc testing for I wasn't aware where the unit test had to be run from to make things work. That being said, we have the same problem in other unit tests. So maybe it's not all that important to fix if people are concerned about this approach.

Reply:
I understand the motivation, I was more considering that there might be a prettier way to have the text inside the file.